src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -497,6 +478,154 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
             }
         }
 
+        // Fill the buffer up to the minimum specified size (or more, if possible).
+        // Returns 0 if EOF on initial read, otherwise throws on EOF.
+        // Returns minSize on success.
+        private int FillBuffer(int minSize)
+        {
+            Debug.Assert(_internalOffset == 0);
+            Debug.Assert(minSize > _internalBufferCount);
+
+            int initialCount = _internalBufferCount;
+            do
+            {
+                int bytes = _sslState.InnerStream.Read(InternalBuffer, _internalBufferCount, InternalBuffer.Length - _internalBufferCount);
+                if (bytes == 0)
+                {
+                    if (_internalBufferCount != initialCount)
+                    {
+                        // We read some bytes, but not as many as we expected, so throw.
+                        throw new IOException(SR.net_io_eof);
+                    }
+
+                    return 0;
+                }
+
+                _internalBufferCount += bytes;
+            } while (_internalBufferCount < minSize);
+
+            return minSize;
+        }
+
+        // Fill the buffer up to the minimum specified size (or more, if possible).
+        // Returns 0 if EOF on initial read, otherwise throws on EOF.
+        // Returns minSize on success.
+        public async Task<int> FillBufferAsync(int minSize)
+        {
+            Debug.Assert(_internalOffset == 0);
+            Debug.Assert(minSize > _internalBufferCount);
+
+            int initialCount = _internalBufferCount;
+            do
+            {
+                int bytes = await _sslState.InnerStream.ReadAsync(InternalBuffer, _internalBufferCount, InternalBuffer.Length - _internalBufferCount, CancellationToken.None).ConfigureAwait(false);
+                if (bytes == 0)
+                {
+                    if (_internalBufferCount != initialCount)
+                    {
+                        // We read some bytes, but not as many as we expected, so throw.
+                        throw new IOException(SR.net_io_eof);
+                    }
+
+                    return 0;
+                }
+
+                _internalBufferCount += bytes;
+            } while (_internalBufferCount < minSize);
+
+            return minSize;
+        }
+
+        private static async void TaskToAsyncProtocolRequest2(Task<int> task, AsyncProtocolRequest asyncRequest)
```

Comment:
> suggestions welcome

If you follow my suggestion at https://github.com/dotnet/corefx/pull/21574/files#r124086892, we can just manually inline it (since the async void isn't needed), and then we don't need to worry about a name :wink: